rwengine/src/render/GameRenderer.cpp

```diff
@@ -573,126 +520,35 @@ void GameRenderer::drawColour(const glm::vec4& colour, glm::vec4 extents) {
-                carlines.push_back(end->position + glm::vec3(0.f, 0.f, 1.f));
-            }
-        }
-    }
-
-    glm::mat4 model;
-    glUniformMatrix4fv(uniModel, 1, GL_FALSE, glm::value_ptr(model));
-    glEnableVertexAttribArray(posAttrib);
-
-    glBindBuffer(GL_ARRAY_BUFFER, debugVBO);
-    glBufferData(GL_ARRAY_BUFFER, sizeof(glm::vec3) * carlines.size(),
-    &(carlines[0]), GL_STREAM_DRAW);
-    glVertexAttribPointer(posAttrib, 3, GL_FLOAT, GL_FALSE, 0, 0);
-
-    float img[] = {1.f, 0.f, 0.f};
-    glTexImage2D(
-        GL_TEXTURE_2D, 0, GL_RGB, 1, 1,
-        0, GL_RGB, GL_FLOAT, img
-    );
-
-    glDrawArrays(GL_LINES, 0, carlines.size());
-
-    glBufferData(GL_ARRAY_BUFFER, sizeof(glm::vec3) * pedlines.size(),
-    &(pedlines[0]), GL_STREAM_DRAW);
-    glVertexAttribPointer(posAttrib, 3, GL_FLOAT, GL_FALSE, 0, 0);
+    renderer->useProgram(ssRectProg.get());
+    renderer->setUniform(ssRectProg.get(), "colour", colour);
+    renderer->setUniform(ssRectProg.get(), "size", glm::vec2{extents.z, extents.w});
+    renderer->setUniform(ssRectProg.get(), "offset", glm::vec2{extents.x, extents.y});
 
-    float img2[] = {0.f, 1.f, 0.f};
-    glTexImage2D(
-        GL_TEXTURE_2D, 0, GL_RGB, 1, 1,
-        0, GL_RGB, GL_FLOAT, img2
-    );
-
-    glDrawArrays(GL_LINES, 0, pedlines.size());
+    Renderer::DrawParameters wdp;
+    wdp.depthMode = DepthMode::OFF;
+    wdp.blendMode = BlendMode::BLEND_ALPHA;
+    wdp.count = ssRectGeom.getCount();
+    wdp.textures = {texture ? texture->getName() : 0};
 
-    pedlines.clear();
-    carlines.clear();
-    glBindVertexArray( 0 );*/
+    renderer->drawArrays(glm::mat4(1.0f), &ssRectDraw, wdp);
 }
 
 void GameRenderer::renderLetterbox() {
-    /// @todo rewrite this render code to use renderer class
-    glUseProgram(ssRectProgram);
-    glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
-    const float cinematicExperienceSize = 0.15f;
-    glUniform2f(ssRectOffset, 0.f, -1.f * (1.f - cinematicExperienceSize));
-    glUniform2f(ssRectSize, 1.f, cinematicExperienceSize);
-
-    glBindTexture(GL_TEXTURE_2D, 0);
-    glUniform1i(ssRectTexture, 0);
-    glUniform4f(ssRectColour, 0.f, 0.f, 0.f, 1.f);
-
-    glBindVertexArray(ssRectDraw.getVAOName());
-    glDrawArrays(GL_TRIANGLE_STRIP, 0, 4);
-
-    glUniform2f(ssRectOffset, 0.f, 1.f * (1.f - cinematicExperienceSize));
-    glUniform2f(ssRectSize, 1.f, cinematicExperienceSize);
+    constexpr float cinematicExperienceSize = 0.15f;
+    renderer->useProgram(ssRectProg.get());
+    renderer->setUniform(ssRectProg.get(), "colour", glm::vec4{0.f, 0.f, 0.f, 1.f});
+    renderer->setUniform(ssRectProg.get(), "size", glm::vec2{1.f, cinematicExperienceSize});
+    renderer->setUniform(ssRectProg.get(), "offset", glm::vec2{0.f,-1.f * (1.f - cinematicExperienceSize)});
+    Renderer::DrawParameters wdp;
+    wdp.depthMode = DepthMode::OFF;
+    wdp.blendMode = BlendMode::BLEND_NONE;
+    wdp.count = ssRectGeom.getCount();
+    wdp.textures = {0};
```

Comment:
nit: generally we do here a lot of dynamic allocations. Can we replace type of `textures` aka `std::vector` with templated `std::array`?

Reply:
That is possible, but it would fix the number of textures we could use at once.

It's a reasonable change to make, as we don't have that many textures to worry about.

Reply:
Objects reusing is also an option, but I guess It'll require even more work.

Reply:
Yes. I prefer using the fixed size array to explicitly set the texture state for the draw.